File: problems/36-digit-dp/001-numbers-with-repeated-digits/java/Solution.java
import java.util.Arrays;

public class Solution {
    private String digits;
    private Integer[][][][] memo;

    public int numDupDigitsAtMostN(int n) {
        digits = String.valueOf(n);
        int length = digits.length();
        memo = new Integer[length][1 << 10][2][2];
        int unique = dfs(0, 0, true, false);
        return n - unique;
    }

    private int dfs(int pos, int mask, boolean tight, boolean started) {
        if (pos == digits.length()) return started ? 1 : 0;
        int t = tight ? 1 : 0;
        int s = started ? 1 : 0;
        if (memo[pos][mask][t][s] != null) return memo[pos][mask][t][s];

        int limit = digits.charAt(pos) - '0';
        int total = 0;
        for (int digit = 0; digit <= limit; digit++) {
            boolean nextTight = tight && digit == limit;
            if (!started && digit == 0) {
                total += dfs(pos + 1, mask, nextTight, false);
            } else if ((mask & (1 << digit)) == 0) {
                total += dfs(pos + 1, mask | (1 << digit), nextTight, true);
            }
        }
        return memo[pos][mask][t][s] = total;
    }
}

File: problems/36-digit-dp/001-numbers-with-repeated-digits/python/solution.py
from functools import lru_cache


def num_dup_digits_at_most_n(n: int) -> int:
    digits = str(n)

    @lru_cache(maxsize=None)
    def dfs(pos: int, mask: int, tight: bool, started: bool) -> int:
        if pos == len(digits):
            return 1 if started else 0

        bound_digit = int(digits[pos])
        limit = bound_digit if tight else 9
        total = 0

        for digit in range(limit + 1):
            next_tight = tight and digit == bound_digit
            if not started and digit == 0:
                total += dfs(pos + 1, mask, next_tight, False)
            elif not (mask & (1 << digit)):
                total += dfs(pos + 1, mask | (1 << digit), next_tight, True)

        return total

    unique = dfs(0, 0, True, False)
    return n - unique
